### PR TITLE
Optimistic concurrency control

### PR DESCRIPTION
**Pull request by cakirmuha**

### CODE REVIEW

**Comment by olivere** [comment]
This seems legitimate. Will put it on the list for the next release.

For self-reference:
- https://github.com/elastic/elasticsearch/blob/7.2/server/src/main/java/org/elasticsearch/index/get/GetResult.java#L243-L246
- https://github.com/elastic/elasticsearch/blob/7.2/server/src/main/java/org/elasticsearch/action/update/UpdateRequestBuilder.java#L153-L175

**Comment by cakirmuha** [comment]
Yes, it is tested and I am using it in my project from my fork as below:

`require github.com/olivere/elastic/v7 v7.0.4`
`replace github.com/olivere/elastic/v7 => github.com/cakirmuha/elastic/v7 v7.0.8`

**Comment by milinykh** [comment]
By the way, `PrimaryTerm` is already present in some of structs and it uses `int64` type. 

Furthermore, there are still more changes to be made in order to support Optimistic Concurrency Control fully (e.g. IndexService, DeleteService, BulkIndexRequest and BulkDeleteRequest).

**Comment by cakirmuha** [comment]
@milinykh  it is ok, I will add, but I couldn't make sure whether `BulkUpdateRequest` uses optimistic control

**Comment by milinykh** [comment]
It would be great! As for optimistic control in Bulk API, please see: https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#bulk-optimistic-concurrency-control

**Comment by olivere** [comment]
Bulk operations use the non-bulk `IndexRequest`, `DeleteRequest` and `UpdateRequest` classes ([code](https://github.com/elastic/elasticsearch/blob/7.2/server/src/main/java/org/elasticsearch/action/bulk/BulkRequestBuilder.java#L96-L110)). And e.g. [`UpdateRequest` has those fields](https://github.com/elastic/elasticsearch/blob/7.2/server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java#L71-L72).

**Comment by cakirmuha** [comment]
@olivere  I am already using `UpdateRequest`, but Bulk API documentation mentions only `delete` and `index` actions in optimistic concurrency part, so I asked it, but you are right and I will add it. Also, I couldn't understand that `UpdateResponse` struct has `GetResult` field in Go client, it is not available in code (<a href ="https://github.com/elastic/elasticsearch/blob/7895dc9915422a33bb6814ec420c3b068aef4410/server/src/main/java/org/elasticsearch/action/update/UpdateResponse.java#L100" >code</a>)

**Comment by olivere** [comment]
@cakirmuha Don't look too much at the API documentation. It's good, but it's high level. Look at the code. That's where the truth lies. For example, [the `toXContent` method that writes the `GetResult`](https://github.com/elastic/elasticsearch/blob/7.3/server/src/main/java/org/elasticsearch/action/update/UpdateResponse.java#L94-L98).

File: bulk_delete_request.go [hunk 1]
```diff
@@ -116,6 +117,18 @@ func (r *BulkDeleteRequest) VersionType(versionType string) *BulkDeleteRequest {
...
+	r.ifSeqNo = &ifSeqNo
+	return r
+}
+
+// IfPrimaryTerm specifies primary term of the last operation.
+func (r *BulkDeleteRequest) IfPrimaryTerm(ifPrimaryTerm int64) *BulkDeleteRequest {
+	r.ifPrimaryTerm = &ifPrimaryTerm
+	return r
+}
+
```
Comment (milinykh) [inline bulk_delete_request.go:133]: It seems that the code below:
```
	command := bulkDeleteRequestCommand{
		"delete": bulkDeleteRequestCommandOp{
			Index:       r.index,
			Type:        r.typ,
			Id:          r.id,
			Routing:     r.routing,
			Parent:      r.parent,
			Version:     r.version,
			VersionType: r.versionType,
		},
	}
```
should be changed as well.

File: bulk_index_request.go [hunk 2]
```diff
@@ -151,6 +153,18 @@ func (r *BulkIndexRequest) RetryOnConflict(retryOnConflict int) *BulkIndexReques
...
+	r.ifSeqNo = &ifSeqNo
+	return r
+}
+
+// IfPrimaryTerm specifies primary term of the last operation.
+func (r *BulkIndexRequest) IfPrimaryTerm(ifPrimaryTerm int64) *BulkIndexRequest {
+	r.ifPrimaryTerm = &ifPrimaryTerm
+	return r
+}
+
```
Comment (milinykh) [inline bulk_index_request.go:169]: It seems that the code below:
```
	indexCommand := bulkIndexRequestCommandOp{
		Index:           r.index,
		Type:            r.typ,
		Id:              r.id,
		Routing:         r.routing,
		Parent:          r.parent,
		Version:         r.version,
		VersionType:     r.versionType,
		RetryOnConflict: r.retryOnConflict,
		Pipeline:        r.pipeline,
	}
```
should be changed as well.

File: bulk_update_request.go [hunk 3]
```diff
@@ -146,6 +148,18 @@ func (r *BulkUpdateRequest) RetryOnConflict(retryOnConflict int) *BulkUpdateRequ
...
+	r.ifSeqNo = &ifSeqNo
+	return r
+}
+
+// IfPrimaryTerm specifies primary term of the last operation.
+func (r *BulkUpdateRequest) IfPrimaryTerm(ifPrimaryTerm int64) *BulkUpdateRequest {
+	r.ifPrimaryTerm = &ifPrimaryTerm
+	return r
+}
+
```
Comment (milinykh) [inline bulk_update_request.go:164]: It seems that the code below:
```
	updateCommand := bulkUpdateRequestCommandOp{
		Index:           r.index,
		Type:            r.typ,
		Id:              r.id,
		Routing:         r.routing,
		Parent:          r.parent,
		Version:         r.version,
		VersionType:     r.versionType,
		RetryOnConflict: r.retryOnConflict,
	}
```
should be changed as well.

**Comment by olivere** [comment]
For reference: Support for v6 will come with 6.2.22 (18311d66845642561a1b8f857407bdd3e2485a5f).

I will probably use that commit to integrate into 7.x.

**Pull request closed without merging**